File: pkg/termite/lib/tabular/ir.go
// Copyright 2025 Antfly, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package tabular provides inference for traditional ML models (tree ensembles,
// linear models, SVMs) using a SIMD-accelerated pure Go engine.
//
// Models are represented using a framework-agnostic intermediate representation
// (IR) that can be produced by converting XGBoost, LightGBM, sklearn, or CatBoost
// models via the termite-convert tool.
package tabular

import "context"

// TaskType describes the prediction task.
type TaskType string

const (
	TaskRegression           TaskType = "regression"
	TaskBinaryClassification TaskType = "binary_classification"
	TaskMulticlass           TaskType = "multiclass"
	TaskRanking              TaskType = "ranking"
)

// ActivationType describes the output activation function.
type ActivationType string

const (
	ActivationIdentity ActivationType = "identity"
	ActivationSigmoid  ActivationType = "sigmoid"
	ActivationSoftmax  ActivationType = "softmax"
	ActivationExp      ActivationType = "exp"
)

// StageType identifies a pipeline stage.
type StageType string

const (
	StageTreeEnsemble StageType = "tree_ensemble"
	StageLinear       StageType = "linear"
	StageSVM          StageType = "svm"
	StageScaler       StageType = "scaler"
	StageImputer      StageType = "imputer"
)

// TabularModel is the top-level IR for a traditional ML model.
type TabularModel struct {
	SchemaVersion int       `json:"schema_version"`
	Metadata      Metadata  `json:"metadata"`
	Pipeline      []Stage   `json:"pipeline"`
	Output        OutputCfg `json:"output"`
}

// Metadata describes the model's provenance and schema.
type Metadata struct {
	Name            string   `json:"name"`
	SourceFramework string   `json:"source_framework"`
	SourceVersion   string   `json:"source_version,omitempty"`
	Task            TaskType `json:"task"`
	NumFeatures     int      `json:"num_features"`
	NumClasses      int      `json:"num_classes,omitempty"`
	FeatureNames    []string `json:"feature_names,omitempty"`
	FeatureTypes    []string `json:"feature_types,omitempty"`
	CreatedAt       string   `json:"created_at,omitempty"`
}

// OutputCfg describes post-processing applied after the model stage.
type OutputCfg struct {
	Activation ActivationType `json:"activation"`
	NumOutputs int            `json:"num_outputs"`
}

// Stage is a single step in the inference pipeline.
// Exactly one of the typed fields will be non-nil.
type Stage struct {
	Type         StageType     `json:"type"`
	TreeEnsemble *TreeEnsemble `json:"tree_ensemble,omitempty"`
	Linear       *LinearModel  `json:"linear,omitempty"`
	SVM          *SVMModel     `json:"svm,omitempty"`
	Scaler       *Scaler       `json:"scaler,omitempty"`
	Imputer      *Imputer      `json:"imputer,omitempty"`
}

// TreeNodes stores tree data in a struct-of-arrays (SoA) layout for
// cache-friendly, SIMD-amenable traversal. All slices are the same length
// (total node count across all trees). Tree boundaries are given by TreeStarts.
type TreeNodes struct {
	FeatureIndex []int32   `json:"feature_index"` // -1 = leaf node
	Threshold    []float64 `json:"threshold"`
	LeftChild    []int32   `json:"left_child"`  // -1 = none
	RightChild   []int32   `json:"right_child"` // -1 = none
	LeafValue    []float64 `json:"leaf_value"`  // non-zero only for leaves
	DefaultLeft  []bool    `json:"default_left"`
	TreeStarts   []int32   `json:"tree_starts"` // index of each tree's root node
}

// TreeAnnotations holds optional optimiser metadata.
type TreeAnnotations struct {
	ThresholdPrecision   []string `json:"threshold_precision,omitempty"`
	DeadLeavesEliminated int      `json:"dead_leaves_eliminated,omitempty"`
	BranchOrder          string   `json:"branch_order,omitempty"`
}

// TreeEnsemble represents a gradient-boosted or random-forest model.
type TreeEnsemble struct {
	Objective   string          `json:"objective"`
	BaseScore   float64         `json:"base_score"`
	NumTrees    int             `json:"num_trees"`
	NumFeatures int             `json:"num_features"`
	MaxDepth    int             `json:"max_depth"`
	Nodes       TreeNodes       `json:"nodes"`
	Annotations TreeAnnotations `json:"annotations,omitempty"`
}

// LinearModel represents a logistic/linear regression.
// Activation is handled by Output.Activation in the pipeline, not here.
type LinearModel struct {
	Weights [][]float64 `json:"weights"` // [num_outputs][num_features]
	Biases  []float64   `json:"biases"`  // [num_outputs]
}

// SVMModel represents a support vector machine.
type SVMModel struct {
	Kernel          string      `json:"kernel"` // rbf, linear, poly, sigmoid
	SupportVectors  [][]float64 `json:"support_vectors"`
	DualCoefficients [][]float64 `json:"dual_coefficients"`
	Intercept       []float64   `json:"intercept"`
	Gamma           float64     `json:"gamma,omitempty"`
	Coef0           float64     `json:"coef0,omitempty"`
	Degree          int         `json:"degree,omitempty"`
}

// Scaler represents feature standardisation / normalisation.
type Scaler struct {
	Method string    `json:"method"` // standard, minmax, robust, maxabs
	Center []float64 `json:"center,omitempty"`
	Scale  []float64 `json:"scale,omitempty"`
	Min    []float64 `json:"min,omitempty"`
	Max    []float64 `json:"max,omitempty"`
}

// Imputer fills missing (NaN) values.
type Imputer struct {
	Strategy   string    `json:"strategy"` // mean, median, most_frequent, constant
	FillValues []float64 `json:"fill_values"`
}

// Predictor is the public interface for tabular model inference.
type Predictor interface {
	// Predict runs inference on a batch of samples.
	// features: [batch_size][num_features]
	// Returns: [batch_size][num_outputs]
	Predict(ctx context.Context, features [][]float32) ([][]float32, error)

	// PredictSingle runs inference on one sample (optimised fast path).
	PredictSingle(features []float32) ([]float32, error)

	// NumFeatures returns the expected input width.
	NumFeatures() int

	// NumOutputs returns the output width.
	NumOutputs() int

	// Metadata returns model metadata.
	ModelMetadata() *Metadata

	// Close releases resources.
	Close() error
}

File: tools/termite-convert/termite_convert/ir.py
# Copyright 2025 Antfly, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""IR types matching the Go tabular model format."""

from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional


@dataclass
class TreeNodes:
    feature_index: list[int]
    threshold: list[float]
    left_child: list[int]
    right_child: list[int]
    leaf_value: list[float]
    default_left: list[bool]
    tree_starts: list[int]


@dataclass
class TreeAnnotations:
    threshold_precision: list[str] = field(default_factory=list)
    dead_leaves_eliminated: int = 0
    branch_order: str = ""


@dataclass
class TreeEnsemble:
    objective: str
    base_score: float
    num_trees: int
    num_features: int
    max_depth: int
    nodes: TreeNodes
    annotations: TreeAnnotations = field(default_factory=TreeAnnotations)


@dataclass
class LinearModel:
    weights: list[list[float]]
    biases: list[float]
    activation: str = "identity"


@dataclass
class Scaler:
    method: str
    center: list[float] = field(default_factory=list)
    scale: list[float] = field(default_factory=list)
    min: list[float] = field(default_factory=list)
    max: list[float] = field(default_factory=list)


@dataclass
class Imputer:
    strategy: str
    fill_values: list[float] = field(default_factory=list)


@dataclass
class SVMModel:
    kernel: str  # rbf, linear, poly, sigmoid
    support_vectors: list[list[float]]
    dual_coefficients: list[list[float]]
    intercept: list[float]
    gamma: float = 0.0
    coef0: float = 0.0
    degree: int = 0


@dataclass
class Stage:
    type: str
    tree_ensemble: Optional[TreeEnsemble] = None
    linear: Optional[LinearModel] = None
    svm: Optional[SVMModel] = None
    scaler: Optional[Scaler] = None
    imputer: Optional[Imputer] = None


@dataclass
class Metadata:
    name: str
    source_framework: str
    task: str
    num_features: int
    source_version: str = ""
    num_classes: int = 0
    feature_names: list[str] = field(default_factory=list)
    feature_types: list[str] = field(default_factory=list)
    created_at: str = ""


@dataclass
class OutputCfg:
    activation: str
    num_outputs: int


@dataclass
class TabularModel:
    schema_version: int
    metadata: Metadata
    pipeline: list[Stage]
    output: OutputCfg

    def to_dict(self) -> dict:
        """Convert to a JSON-serializable dict, omitting empty optional fields."""
        d = _clean_dict(asdict(self))
        return d

    def save(self, path: Path):
        """Write tabular_model.json to the given directory."""
        path.mkdir(parents=True, exist_ok=True)
        model_file = path / "tabular_model.json"
        with open(model_file, "w") as f:
            json.dump(self.to_dict(), f, indent=2)

    def save_manifest(self, path: Path):
        """Write model_manifest.json for Termite discovery."""
        manifest = {
            "model_type": "predictor",
            "capabilities": _capabilities_for_model(self),
            "metadata": {
                "name": self.metadata.name,
                "source_framework": self.metadata.source_framework,
                "task": self.metadata.task,
            },
        }
        manifest_file = path / "model_manifest.json"
        with open(manifest_file, "w") as f:
            json.dump(manifest, f, indent=2)


def _capabilities_for_model(m: TabularModel) -> list[str]:
    caps = ["tabular"]
    for stage in m.pipeline:
        if stage.type == "tree_ensemble":
            caps.append("tree_ensemble")
        elif stage.type == "linear":
            caps.append("linear_model")
        elif stage.type == "svm":
            caps.append("svm")
    return caps


def _clean_dict(d):
    """Recursively remove None values and empty optional fields."""
    if isinstance(d, dict):
        cleaned = {}
        for k, v in d.items():
            v = _clean_dict(v)
            # Skip None and empty strings/lists for optional fields
            if v is None:
                continue
            if isinstance(v, str) and v == "" and k in (
                "source_version", "created_at", "branch_order",
            ):
                continue
            if isinstance(v, list) and len(v) == 0 and k in (
                "feature_names", "feature_types", "threshold_precision",
                "center", "scale", "min", "max", "fill_values",
            ):
                continue
            if isinstance(v, int) and v == 0 and k in (
                "num_classes", "dead_leaves_eliminated",
            ):
                continue
            cleaned[k] = v
        return cleaned
    elif isinstance(d, list):
        return [_clean_dict(item) for item in d]
    return d
